### ISSUE TITLE
Two request can get deduplicated if same query and vars but different headers

### ISSUE THREAD

**Issue by robermorales** [apollographql/apollo-link#920]
Two queries can get wrongly deduplicated if they are requesting the same query and vars to the same host, but with different HTTP headers. Perhaps we can include all or some request Headers into the `getKey` of the operation type, so that case is well covered and never the second request got hidden with the result of the first one.

https://github.com/apollographql/apollo-link/blob/089f71be5ea6a2ccd465751a500b16b97dbc23aa/packages/apollo-link/src/linkUtils.ts#L127

I do not know even how to implement this, but first: is this idea logic, reliable, and feasible??

Thanks,

**Comment by JoviDeCroock** [comment]
Could you provide a use case for this, can't really see any case for this yet. That's probably just me

**Comment by robermorales** [comment]
I reached this situation in my current project. Yes, it is about auth headers, but could be another header.

Neither the HTTP RFC or the graphql specification says nothing against doing several requests at the same time with different headers.

Actually, the headers I see that get wrongly deduped are targeting the nginx proxy between the graphql client and the server, they have nothing to do with graphql, but with other layers of the communication.

Anyway, if we choose carefully the implementation, I think that the change could be harmless for every other use case.

**Comment by JoviDeCroock** [comment]
Yeah was not doubting your need for it or anything, just trying to see if it was a really specific scenario or not, no worries. Let's look at it, I think it can imply some issues though. Not 100% sure

**Comment by robermorales** [comment]
Cool!

Let me know if I can help with the implementation.

Thanks!!!

**Comment by JoviDeCroock** [comment]
In the meanwhile you can probably use `context: { forceFetch: true }` which forces the dedup to be skipped for the Query you give it to.

**Comment by robermorales** [comment]
> In the meanwhile you can probably use `context: { forceFetch: true }` which forces the dedup to be skipped for the Query you give it to.

Thanks!!!

**Comment by Druotic** [comment]
@robermorales I'm not sure if you're referring specifically to the dedup link or if you are using the higher level client. If you are using apollo-link-dedup then https://github.com/apollographql/apollo-link/pull/957 may address your concerns, as well.

**Comment by robermorales** [comment]
>  I'm not sure if you're referring specifically to the dedup link or if you are using the higher level client. If you are using apollo-link-dedup then #957 may address your concerns, as well.

looks good, @Druotic
